controllers/virtualmachinegrouppublishrequest/virtualmachinegrouppublishrequest_controller.go

```diff
@@ -95,63 +110,288 @@ func NewReconciler(
+		return r.removeAfterTTL(ctx)
+	}
+
+	if err := r.verifySpec(vmGroupPublishReq); err != nil {
+		// in the rare case that webhooks were down when request was created
+		// set condition to failure
+		conditions.MarkError(vmGroupPublishReq,
+			vmopv1.VirtualMachineGroupPublishRequestConditionComplete,
+			invalidSpecErrorMsg,
+			err)
+		return pkgerr.NoRequeueError{Message: invalidSpecErrorMsg}
+	}
+
+	if vmGroupPublishReq.Status.StartTime.IsZero() {
+		vmGroupPublishReq.Status.StartTime = metav1.NewTime(time.Now())
+	}
+
+	return r.reconcileStatus(ctx)
+}
+
+func (r *Reconciler) verifySpec(vmGroupPublishReq *vmopv1.VirtualMachineGroupPublishRequest) error {
+	var errs []error
+
+	if vmGroupPublishReq.Spec.Source == "" {
+		errs = append(errs, fmt.Errorf(undefinedSpecErrorMsg, "source"))
+	}
+	if vmGroupPublishReq.Spec.Target == "" {
+		errs = append(errs, fmt.Errorf(undefinedSpecErrorMsg, "target"))
+	}
+	if len(vmGroupPublishReq.Spec.VirtualMachines) == 0 {
+		errs = append(errs, fmt.Errorf(undefinedSpecErrorMsg, "virtualMachines"))
+	}
+
+	return errors.Join(errs...)
+}
+
+func (r *Reconciler) reconcileStatus(ctx *pkgctx.VirtualMachineGroupPublishRequestContext) error {
+
+	createdRequestsMap, pendingRequestsSet, completedRequestsSet, err := r.getVMPublishRequests(ctx)
+	if err != nil {
+		return err
+	}
+
+	ctx.VMGroupPublishRequest.Status.Images = r.getSortedImageStatus(createdRequestsMap)
+
+	desiredRequestsSet := sets.Set[string]{}
+	for _, vm := range ctx.VMGroupPublishRequest.Spec.VirtualMachines {
+		desiredRequestsSet.Insert(vm)
+	}
+
+	if err := r.createNewVMPublishRequests(ctx, createdRequestsMap, desiredRequestsSet); err != nil {
+		return err
+	}
+
+	if completed := r.isCompleted(ctx.VMGroupPublishRequest, pendingRequestsSet, completedRequestsSet, desiredRequestsSet); !completed {
+		return pkgerr.RequeueError{After: pkgmgr.DefaultSyncPeriod}
+	}
+
+	return pkgerr.RequeueError{}
+}
+
+func (r *Reconciler) getVMPublishRequests(
+	ctx *pkgctx.VirtualMachineGroupPublishRequestContext) (
+	map[string]vmopv1.VirtualMachinePublishRequest,
+	sets.Set[string],
+	sets.Set[string],
+	error) {
+
+	requests := &vmopv1.VirtualMachinePublishRequestList{}
+	if err := r.List(ctx, requests, client.InNamespace(ctx.VMGroupPublishRequest.Namespace)); err != nil {
+		ctx.Logger.Error(err, "failed to list VirtualMachinePublishRequest")
+		return nil, nil, nil, err
+	}
+
+	createdRequestsMap := make(map[string]vmopv1.VirtualMachinePublishRequest)
+	pendingRequestsSet := sets.Set[string]{}
+	completedRequestsSet := sets.Set[string]{}
+	for _, request := range requests.Items {
+		owned, err := controllerutil.HasOwnerReference(request.OwnerReferences, ctx.VMGroupPublishRequest, r.Scheme())
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		if owned {
+			createdRequestsMap[request.Spec.Source.Name] = request
+			if conditions.IsTrue(&request, vmopv1.VirtualMachinePublishRequestConditionComplete) && request.Status.Ready {
```

Comment:
Is there a reason that just checking `request.Status.Ready` isn't good enough?

Reply:
no, I just saw they were set at the same time  in the VM publish controller